Ibf dev

lbologna opened the pull request:
Hello @wvangeit,

I have integrated the current version of the code that you uploaded a few weeks ago with what we have developed on a parallel branch for our needs. 
I tested with both python3 and python2 and the results seem to be the same as the ones obtained with the code in the 'master' branch (unless specific options affecting the results are used).

I give some explanation below, about the changes/integrations that have been done, but please let me know if some more information is required.

Thank you.

#########################
The option "stim_info" has been maintained for compatibility and extended. It can now be:
1) a dictionary (as before)
2) a list of dictionaries (one per file or a single one to be used for every file)
(when "stim_info" is set, stimulus amplitudes are extracted from the traces (this may cause loss of precision if the stimulus signal is noisy))

#########################
The option "stim_feats" has been maintained as it uses different keys (the same as in the metadata files) for some values. As for the "stim_info" option it can be:

1) a dictionary
2) a list of dictionaries (one per file, or a single one to be used for every file)

Conceptually, it is the same as "stim_info". If both values are found in the configuration file, an exception is raised.

#########################
In the configuration file, a new option has been introduced named "zero_to_nan".
It allows to convert to "nan" or to "stim_end" (stimulus' end time) the feature values that equal to zero. It consists in a dictionary having the following form:

    "zero_to_nan":{                                                            
        "flag": true,                                                          
            "value": "nan",                                                        
            "mean_features_no_zeros":[                                             
                "inv_first_ISI", "inv_second_ISI",                                 
                "inv_third_ISI", "inv_fourth_ISI", 
		"inv_fifth_ISI", "inv_last_ISI", 		  
                "time_to_first_spike", 	
		"time_to_last_spike"        
            ]                                                                      
        },
if "flag" is set to 'false' this option is not applied

the 'value' key can have value either 'nan' or 'stim_end'
'mean_features_no_zeros' indicate the features whose zero value has to be converted to 'nan' or 'stim_end'


#########################
The 'v_corr' option is now managed in such a way that it can be a numeric value, a one-element list or a list.
In the first two cases the given value will be added to all the files. In the latter case, a value must be specified for each file.

#########################
The data dictionary is now filled through the 'common.manageDicts.fill_dict_single_trace' function, which can be used for all data types (axon, igor, ...)

#########################
The function 'stim_feats_from_meta' has been moved from 'axon.py' to the 'common.py' module, given that it is common to and can be used with all data type (axon, igor, ...)

#########################
The optional configuration parameter 'target_unit' has been added, that specifies the units in which the 'target' values are specified. If this parameter is present in 'options', the values of 'target' are converted to 'nA' according to the indicated 'target_unit'. Otherwise, they are left unchanged.

#########################
New option added: "zero_std", if set to through, it will consider also values having std=0 and not only values having either std > 0.0 or mean = 0.0

#########################
The optional field "exclude_unit" has been added. If present, it must have the same length as "exclude" and will convert the given values to 'nA' according to the indicated units.

#########################
A new option was introduced in order to print individual feature values in a table stored in a .txt file named 'all_feature_table.txt' and saved in the top level of the results folder.
Use is as follows:

        "print_table": {
            "flag": true,
            "num_events": 5
        }

where 'flag' indicates whether this functionality has to be used, and 'num_events' indicates how many events have to be saved in case a multi-event (e.g. AP_width) feature has been extracted

the file 'tabletools.py' has been added to this purpose

#########################
Three config .json files have been added for testing in the 'tests/configs' folder

#########################
Added configuration files for the 'print_table' option:

tabletools_config.json 
features_default.json

and a for metadata file management:

metadata_template.json

wvangeit commented:
Hi @lbologna ,
Could you check the test that are failing?
Thanks.

anilbey commented on bluepyefe/formats/common.py:
```diff
@@ -0,0 +1,563 @@
+            with open(filename_meta) as f:
+                data = json.load(f)
+
+                if "animal_species" not in data or "animal_species" == None:
```
There's a mistake here.
`"animal_species" == None` will always evaluate `False`.
It is same as writing:
`if "animal_species" not in data or False:`
Which can be simplified to
`if "animal_species not in data:"`

anilbey replied:
You probably wanted to write `data["animal_species"] == None`, I think

wvangeit replied:
@lbologna, was this resolved?

lbologna replied:
Hello @wvangeit,

yes, this was modified for all field in the dictionary.
Thanks

lbologna commented:
Hello @wvangeit, @anilbey,

thanks for the request changes, I fixed the error.

As per the tests which are failing, it seems that one configuration file (.json) that I created and inserted in the repository cannot be found.

The file is this one:

https://github.com/lbologna/BluePyEfe/blob/ibf_dev/bluepyefe/formats/configs/metadata_template.json

and the error is this one:

IOError: [Errno 2] No such file or directory: '/home/travis/build/BlueBrain/BluePyEfe/.tox/py27/lib/python2.7/site-packages/bluepyefe/formats/configs/metadata_template.json'

When I install BluePyEfe in a local python virtual environment everything runs well.


In order to use this and other config files, I added some lines in the setup.py file:

https://github.com/BlueBrain/BluePyEfe/pull/10/files#diff-2eeaed663bd0d25b7e608891384b7298R62

I noticed though that, locally, the path is the following:
venvp3_test/lib/python3.8/site-packages/bluepyefe-0.2.24-py3.8.egg/bluepyefe/formats/configs/

so it contains the bluepyefe-0.2.24-py3.8.egg whose presence I am not able to verify from the travis log.

Do you have any suggestion/feedback on this?

Thank you!

> Hi @lbologna ,
> Could you check the test that are failing?
> Thanks.

wvangeit commented:
@lbologna did you try to run tox locally on your machine?

lbologna commented:
> @lbologna did you try to run tox locally on your machine?

Hello @wvangeit,

I did some changes and tested the code locally with tox on a different branch.
You can see the test results here:
https://travis-ci.org/github/lbologna/BluePyEfe/builds/691002268

It seems everything is working except for the python2.7 test, which gives an error in the plt_features_dist function
https://github.com/lbologna/BluePyEfe/blob/master/bluepyefe/extractor.py#L1220
when an array containing a nan value is provided:
https://travis-ci.org/github/lbologna/BluePyEfe/jobs/691002269#L345

I am not sure why this error is not raised with python3 but I was wandering whether it would make sense, for the logic of the program, to replace the following instruction:
**all(numpy.isnan(feat)) is False**
with the following one:
**any(numpy.isnan(feat)) is False**
in this line:
https://github.com/lbologna/BluePyEfe/blob/master/bluepyefe/extractor.py#L1262
(if this makes sense, this should solve the issue).

Thanks.

wvangeit commented:
@lbologna It seems there is still a problem with the FileNotFoundError. When I run the tests on my own computer I also get this error. Maybe you can try to start from a clean clone on your machine and see if you can reproduce it?

wvangeit commented:
Ok, now the tests seem to work apart from the nan problem.
@DrTaDa do you have any comments about the above suggestions to catch the NaNs.
I guess it will imply that 1 failed feature might prevent the others from being plotted?

lbologna commented:
> @lbologna It seems there is still a problem with the FileNotFoundError. When I run the tests on my own computer I also get this error. Maybe you can try to start from a clean clone on your machine and see if you can reproduce it?

Hello @wvangeit,

Thanks for checking.
You still had the same error because I had not pushed the last changes on the branch of the pull request (ibf_dev) but was developing on another branch in my fork (ibf_dev_test_tox) given that I was still having one failure with the python2.7.

I merged my two branches now, so you should see the results of the latest test:
https://travis-ci.org/github/BlueBrain/BluePyEfe/builds/692147674
This is the same result that I obtain locally running tox.

Thanks.

DrTaDa commented:
@wvangeit @lbologna I would recommend not to change all by any as it happens quite often that a feature value is nan which would results in most of the plots being blank. Instead we could  remove the nans form the array before plotting, eg:

`feat = feat[numpy.logical_not(numpy.isnan(feat))]`

lbologna commented:
> @wvangeit @lbologna I would recommend not to change all by any as it happens quite often that a feature value is nan which would results in most of the plots being blank. Instead we could remove the nans form the array before plotting, eg:
> 
> `feat = feat[numpy.logical_not(numpy.isnan(feat))]`

Thanks @DrTaDa for your feedback. It's definitely fine with us to remove the nan before plotting.
If it's fine with you, I will change the code in may branch and test again before pushing to github.
Thanks.

wvangeit commented:
@lbologna yes, that sounds good
